### DISCUSSION TITLE
Deserializing when iterating over children()

### DISCUSSION THREAD

**Discussion by ftpronk** [biojppm/rapidyaml#372] [General]
I'm trying to extract a number of variables stored in a yaml file.  As the number of values/names is unknown beforehand, I'm transversing through the yaml tree.  (In this case, it happens to be a flat tree)

Is there a way to deserialize a child node to a c++ type?  I've managed with a direct tree['key'] access, but not from ryml::Node type.
This is the best I've come up with, but it's a bit clunky..

```
string buffer = readFile(argv[1]);
ryml::Tree tree = ryml::parse_in_place(ryml::to_substr(buffer));
auto root_ref = tree.rootref();

unordered_map<string, double> remap;

for (auto child: root_ref.children()) {
  if (child.has_key() && child.has_val()) {
    string key;
    ryml::from_chars(child.key(), &key);
    root_ref[child.key()] >> remap[key];
  }
}
```

As a bonus, is there a simpler way to convert a csubstr to a std::string?

**Comment by biojppm** [comment]
Sorry for the late reply, missed this notification.

There are several different ways to do that, and they are documented in the quickstart, so have a look over there. Specifically, [in these sample functions](https://github.com/biojppm/rapidyaml/blob/deb42ea7c5a36554ce3d0f745fca6cfaabd3ae67/samples/quickstart.cpp#L70:73):
```c++
void sample_user_scalar_types();    ///< serialize/deserialize scalar (leaf/string) types
void sample_user_container_types(); ///< serialize/deserialize container (map or seq) types
void sample_std_types();            ///< serialize/deserialize STL containers
```

A couple of notes about your code.
```c++
    root_ref[child.key()] >> remap[key];
// Don't do this, you are doing a double lookup.
// root_ref[child.key()] resolves to child, so just do the following instead:
    child >> remap[key];
```

Now to get the key, use the `ryml::key()` tag type:
```c++
child >> ryml::key(key);
```

And reuse the key string to save you n allocations/deallocations:
```c++
std::string key;
for (auto child: root_ref.children()) {
  if (child.has_key() && child.has_val()) {
    child >> ryml::key(key);
    child >> remap[key];
  }
}
```

As for deserializing into the container, after reading the quickstart, you should look at the [c4/yml/std/map.hpp](https://github.com/biojppm/rapidyaml/blob/master/src/c4/yml/std/map.hpp) header providing interop with `std::map`  which is similar to the `std::unordered_map` you are using.

**Reply by ftpronk** [reply]
Thank you for coming back to me.  I'll try that out, and dig through the links you added.
